File: CSharp/Concentus.Silk/Structs/silk_encoder.cs
using Concentus.Common.CPlusPlus;
using Concentus.Common;
using Concentus.Silk.Enums;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Concentus.Silk.Structs
{
    /// <summary>
    /// Encoder Super Struct
    /// </summary>
    public class silk_encoder
    {
        public readonly Pointer<silk_encoder_state_fix> state_Fxx = Pointer.Malloc<silk_encoder_state_fix>(SilkConstants.ENCODER_NUM_CHANNELS);
        public readonly stereo_enc_state sStereo = new stereo_enc_state();
        public int nBitsUsedLBRR = 0;
        public int nBitsExceeded = 0;
        public int nChannelsAPI = 0;
        public int nChannelsInternal = 0;
        public int nPrevChannelsInternal = 0;
        public int timeSinceSwitchAllowed_ms = 0;
        public int allowBandwidthSwitch = 0;
        public int prev_decode_only_middle = 0;

        public silk_encoder()
        {
            for (int c = 0; c < SilkConstants.ENCODER_NUM_CHANNELS; c++)
            {
                state_Fxx[c] = new silk_encoder_state_fix();
            }
        }

        public void Reset()
        {
            for (int c = 0; c < SilkConstants.ENCODER_NUM_CHANNELS; c++)
            {
                state_Fxx[c].Reset();
            }

            sStereo.Reset();
            nBitsUsedLBRR = 0;
            nBitsExceeded = 0;
            nChannelsAPI = 0;
            nChannelsInternal = 0;
            nPrevChannelsInternal = 0;
            timeSinceSwitchAllowed_ms = 0;
            allowBandwidthSwitch = 0;
            prev_decode_only_middle = 0;
        }

        /// <summary>
        /// Initialize Silk Encoder state
        /// </summary>
        /// <param name="psEnc">I/O  Pointer to Silk FIX encoder state</param>
        /// <param name="arch">I    Run-time architecture</param>
        /// <returns></returns>
        public static int silk_init_encoder(silk_encoder_state_fix psEnc, int arch)
        {
            int ret = 0;

            // Clear the entire encoder state
            psEnc.Reset();

            psEnc.sCmn.arch = arch;

            psEnc.sCmn.variable_HP_smth1_Q15 = Inlines.silk_LSHIFT(Inlines.silk_lin2log(Inlines.SILK_FIX_CONST(TuningParameters.VARIABLE_HP_MIN_CUTOFF_HZ, 16)) - (16 << 7), 8);
            psEnc.sCmn.variable_HP_smth2_Q15 = psEnc.sCmn.variable_HP_smth1_Q15;

            // Used to deactivate LSF interpolation, pitch prediction
            psEnc.sCmn.first_frame_after_reset = 1;

            // Initialize Silk VAD
            ret += VAD.silk_VAD_Init(psEnc.sCmn.sVAD);

            return ret;
        }
    }
}

File: Go/opus/silk_encoder.go
package opus

type SilkEncoder struct {
	state_Fxx                 []*SilkChannelEncoder
	sStereo                   *StereoEncodeState
	nBitsUsedLBRR             int
	nBitsExceeded             int
	nChannelsAPI              int
	nChannelsInternal         int
	nPrevChannelsInternal     int
	timeSinceSwitchAllowed_ms int
	allowBandwidthSwitch      int
	prev_decode_only_middle   int
}

func NewSilkEncoder() SilkEncoder {
	enc := SilkEncoder{}
	enc.sStereo = NewStereoEncodeState()
	enc.state_Fxx = make([]*SilkChannelEncoder, ENCODER_NUM_CHANNELS)
	for i := 0; i < ENCODER_NUM_CHANNELS; i++ {
		enc.state_Fxx[i] = NewSilkChannelEncoder()
	}
	return enc
}

func (enc *SilkEncoder) Reset() {
	for c := 0; c < ENCODER_NUM_CHANNELS; c++ {
		enc.state_Fxx[c].Reset()
	}
	enc.sStereo.Reset()
	enc.nBitsUsedLBRR = 0
	enc.nBitsExceeded = 0
	enc.nChannelsAPI = 0
	enc.nChannelsInternal = 0
	enc.nPrevChannelsInternal = 0
	enc.timeSinceSwitchAllowed_ms = 0
	enc.allowBandwidthSwitch = 0
	enc.prev_decode_only_middle = 0
}

func silk_init_encoder(psEnc *SilkChannelEncoder) int {
	ret := 0
	psEnc.Reset()
	value := int(float64(TuningParameters.VARIABLE_HP_MIN_CUTOFF_HZ)*float64(1<<16) + 0.5)
	logVal := silk_lin2log(value)
	psEnc.variable_HP_smth1_Q15 = int(silk_LSHIFT(logVal-(16<<7), 8))
	psEnc.variable_HP_smth2_Q15 = psEnc.variable_HP_smth1_Q15
	psEnc.first_frame_after_reset = 1
	ret += silk_VAD_Init(psEnc.sVAD)
	return ret
}
